ballista/rust/scheduler/src/flight_sql.rs

```diff
@@ -0,0 +1,460 @@
+            .query_stage_event_loop
+            .get_sender()
+            .map_err(|e| {
+                Status::internal(format!(
+                    "Could not get query stage event sender due to: {}",
+                    e
+                ))
+            })?;
+
+        query_stage_event_sender
+            .post_event(QueryStageSchedulerEvent::JobQueued {
+                job_id: job_id.clone(),
+                session_id: ctx.session_id().clone(),
+                session_ctx: ctx,
+                plan: Box::new(plan.clone()),
+            })
+            .await
+            .map_err(|e| {
+                let msg =
+                    format!("Failed to send JobQueued event for {}: {:?}", job_id, e);
+                error!("{}", msg);
+                Status::internal(msg)
+            })?;
+
+        // let handle = Uuid::new_v4();
+        // let mut statements = self.statements.try_lock()
+        //     .map_err(|e| Status::internal(format!("Error locking statements: {}", e)))?;
+        // statements.insert(handle, plan.clone());
+
+        // poll for job completion
+        let mut num_rows = 0;
+        let mut num_bytes = 0;
+        let fieps = loop {
+            sleep(Duration::from_millis(100)).await;
+            let status = self
+                .server
+                .state
+                .task_manager
+                .get_job_status(&job_id)
+                .await
+                .map_err(|e| {
+                    let msg = format!("Error getting status for job {}: {:?}", job_id, e);
+                    error!("{}", msg);
+                    Status::internal(msg)
+                })?;
+            let status: JobStatus = match status {
+                Some(status) => status,
+                None => {
+                    let msg = format!("Error getting status for job {}!", job_id);
+                    error!("{}", msg);
+                    Err(Status::internal(msg))?
+                }
+            };
+            let status: job_status::Status = match status.status {
+                Some(status) => status,
+                None => {
+                    let msg = format!("Error getting status for job {}!", job_id);
+                    error!("{}", msg);
+                    Err(Status::internal(msg))?
+                }
+            };
+            let completed = match status {
+                job_status::Status::Queued(_) => continue,
+                job_status::Status::Running(_) => continue,
+                job_status::Status::Failed(e) => Err(Status::internal(format!(
+                    "Error building plan: {}",
+                    e.error
+                )))?,
+                job_status::Status::Completed(comp) => comp,
+            };
+            let mut fieps: Vec<_> = vec![];
+            for loc in completed.partition_location.iter() {
+                let fetch = if let Some(ref id) = loc.partition_id {
+                    let fetch = protobuf::FetchPartition {
```

Comment:
I built this directly using the prost generated structs, and it was very painful since all this lacked autocomplete in my IDE. It looked like it was also possible to create Ballista/DataFusion native versions of these structs then `From` or `Into` to convert to prost? @andygrove or anyone who knows, I'd love advice on how to do this better.

Reply:
Most of these prost structs have a corresponding struct that is nicer to use (along with conversions from/to). For this case you can convert the prost struct into a `ballista_core::serde::scheduler::PartitionLocation` which is nicer to use (it will unwrap all the `Option` fields which protobuf insists on but are not really optional).